Caboodle/Magnetometer/Magnetometer.shared.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.Caboodle
+{
+    public static partial class Magnetometer
+    {
+        public static bool IsMonitoring =>
+               MonitorCTS != null && !MonitorCTS.IsCancellationRequested;
+
+        public static void Start(SensorSpeed sensorSpeed, Action<MagnetometerData> handler)
+        {
+            if (handler == null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+
+            PreMonitorValidation();
+            CreateToken();
+            PlatformStart(sensorSpeed, handler);
+        }
+
+        public static void Stop()
+        {
+            if (MonitorCTS == null)
+                return;
+
+            if (!MonitorCTS.Token.CanBeCanceled || MonitorCTS.Token.IsCancellationRequested)
+                return;
+
+            MonitorCTS.Cancel();
+        }
+
+        internal static CancellationTokenSource MonitorCTS { get; set; }
+
+        internal static void PreMonitorValidation()
+        {
+            if (!IsSupported)
+            {
+                throw new FeatureNotSupportedException();
+            }
+
+            if (IsMonitoring)
+            {
+                throw new InvalidOperationException("Magnetometer is already being monitored. Please stop to start a new session.");
```

Comment:
I am just looking at the API and I see _why_ this may be a throw, but then why do we use delegates? What was the reason for the choice over events? .NET is very event-y so from my side of UWP/WinForms, the delegate approach is a bit foreign.

The only possibility for delegates is that you can have multiple event streams with different speeds. Is this something that the platforms actually support? Even so, is this something we want to encourage?

What is wrong with:

```csharp
static class Magnetometer {
    static event EventHandler<Data> Updated;
    static void Start(EventSpeed speed);
    static void Stop();
}
```

Reply:
Another thing that I "feel" with events vs delegates is that delegates are more of a demand: "do this here, or under this condition". Events are more of: "when X happens I want to be notified so I can do this"

Reply:
Let's discuss today after the call :)

Reply:
Moving this to a new issue to contain the discussions: https://github.com/xamarin/Caboodle/issues/142